packages/zoltan2/src/algorithms/partition/Zoltan2_AlgMultiJagged.hpp

```diff
@@ -54,7037 +54,8852 @@
+   *  \param inBuffer         DOCWORK: Documentation
+   *  \param inoutBuffer      DOCWORK: Documentation
+   */
+  void reduce( const Ordinal count, const T inBuffer[], T inoutBuffer[]) const {
+    for(Ordinal i = 0; i < count; i++) {
+      if(Z2_ABS(inBuffer[i]) > epsilon) {
+        inoutBuffer[i] = inBuffer[i];
+      }
     }
+  }
 };
+
 } // namespace Teuchos
 
 namespace Zoltan2{
 
-/*! \brief Allocates memory for the given size.
- *
- */
-template <typename T>
-T *allocMemory(size_t size){
-    if (size > 0){
-        T * a = new T[size];
-        if (a == NULL) {
-            throw  "cannot allocate memory";
-        }
-        return a;
-    }
-    else {
-        return NULL;
-    }
-}
-
-/*! \brief Frees the given array.
- *
- */
-template <typename T>
-void freeArray(T *&array){
-    if(array != NULL){
-        delete [] array;
-        array = NULL;
-    }
-}
-
-
 /*! \brief Class for sorting items with multiple values.
- * First sorting with respect to val[0], then val[1] then ... val[count-1].
- * The last tie breaking is done with index values.
- * Used for task mapping partitioning where the points on a cut line needs to be
- * distributed consistently.
- *
+ *  First sorting with respect to val[0], then val[1] then ... val[count-1].
+ *  The last tie breaking is done with index values.
+ *  Used for task mapping partitioning where the points on a cut line needs to
+ *  be distributed consistently.
  */
 template <typename IT, typename CT, typename WT>
 class uMultiSortItem
 {
 public:
-    //TODO: Why volatile?
-    //no idea, another intel compiler faiulure.
-    volatile IT index;
-    volatile CT count;
-    //unsigned int val;
-    volatile WT *val;
-    volatile WT _EPSILON;
-
-    uMultiSortItem(){
-        this->index = 0;
-        this->count = 0;
-        this->val = NULL;
-        this->_EPSILON = std::numeric_limits<WT>::epsilon() * 100;
-    }
-
+  // TODO: Why volatile?
+  // no idea, another intel compiler failure.
+  volatile IT index;
+  volatile CT count;
+  volatile WT *val;
+  volatile WT epsilon;
+
+  uMultiSortItem() {
+    this->index = 0;
+    this->count = 0;
+    this->val = NULL;
+    this->epsilon = std::numeric_limits<WT>::epsilon() * 100;
+  }
 
-    uMultiSortItem(IT index_ ,CT count_, WT *vals_){
-        this->index = index_;
-        this->count = count_;
-        this->val = vals_;
-        this->_EPSILON = std::numeric_limits<WT>::epsilon() * 100;
-    }
+  // TODO: Document these methods?
+  uMultiSortItem(IT index_ ,CT count_, WT *vals_) {
+    this->index = index_;
+    this->count = count_;
+    this->val = vals_;
+    this->epsilon = std::numeric_limits<WT>::epsilon() * 100;
+  }
 
-    uMultiSortItem( const uMultiSortItem<IT,CT,WT>& other ){
-        this->index = other.index;
-        this->count = other.count;
-        this->val = other.val;
-        this->_EPSILON = other._EPSILON;
-    }
+  uMultiSortItem( const uMultiSortItem<IT,CT,WT>& other ) {
```

Comment:
You don't need to provide any of these copy constructors or assignment operators.  The [Rule of Zero](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-zero) will provide correct defaults here.